fec/models.py

```diff
@@ -45,21 +45,128 @@ class Committee(models.Model):
+
+    def first_names(self):
+
+        return self.fixes(self.contributor_name.split(',')[1]).split()
+
+    def last_name(self):
+
+        return self.fixes(self.contributor_name.split(',')[0])
+
+    def fixes(self, string_to_fix):
+
+        for fix in self.f:
+            string_to_fix = string_to_fix.replace(fix, '')
+
+        return string_to_fix
+
+    def get_nicknames(self, name):
+        '''
+        Make 2 data structures - "lines" is a dictionary with given name as key,
+        and "names" which is all names as a list of lists with name and index number
+        '''
+        names, lines = self.get_names_data('./fec/data/names.txt')
+
+        # Search for all names that match first_name
+        all_names = self.nickname_search(name, names)
+
+        if all_names is None:
+            return [name]
+
+        names = [
+            lines[all_names[index][1]][0] for index, _ in enumerate(all_names)
+        ]
+        return names
+
+    @staticmethod
+    def get_names_data(file_path):
+        lines = []
+        with open(file_path) as file:
+            for index, line in enumerate(file.readlines()):
+                lines.append(line.strip('\n').split(','))
+                names = [[name, index] for name in line.split(',')]
+                names.sort()
+            names.sort()
+            return names, lines
+
+    @staticmethod
+    def nickname_search(first_name, names):
+        """ Find and return the index of key in sequence names  for first_name"""
+        lb = 0
+        ub = len(names)
+
+        while True:
+            if lb == ub:  # If region of interest (ROI) becomes empty
+                return None
+            # Next probe should be in the middle of the ROI
+            mid_index = (lb + ub) // 2
+            # Fetch the item at that position
+            item_at_mid = names[mid_index][0]
+            # How does the probed item compare to the target?
+            if item_at_mid == first_name:
+                upper_mid_index = mid_index
+                lower_mid_index = mid_index
+                while names[upper_mid_index + 1][0] == first_name:
+                    upper_mid_index = upper_mid_index + 1
+                while names[lower_mid_index - 1][0] == first_name:
+                    lower_mid_index = lower_mid_index - 1
+                return names[lower_mid_index:upper_mid_index + 1]  # Found it!
+            if item_at_mid < first_name:
+                lb = mid_index + 1  # Use upper half of ROI next time
+            else:
+                ub = mid_index  # Use lower half of ROI next time
+
+    def regex_name(self):
+
+        regex_fixes = '({0})?'.format('|'.join([f + '(.)?' for f in self.f]))
+        try:
+            first_name, *rest_of_first = self.first_names()
+            return regex_fixes + self.last_name() + regex_fixes + \
+                ', ({0})'.format('|'.join(self.get_nicknames(first_name))) + \
+                '({0})?'.format('|'.join(rest_of_first)) + regex_fixes
```

Comment:
Phew, regex can be brutal (but so useful!) I could imagine turning this function into a whole class, to abstract out some of the repeated code. Like maybe it would have a method like this:

```python
def one_of(options): 
  '({0})'.format('|'.join(options))
```

given that I see that pattern pop up a couple times. On the other hand, you end up not having to change them often, then leaving regular expressions as a well-tested black box is often the right call!

For now, I'd be down to leave as is, but consider refactoring if we end up having to change it a lot.